src/ObjectUtils/PdbFileLlvm.cpp

```diff
@@ -214,6 +241,117 @@ PdbFileLlvm::PdbFileLlvm(std::filesystem::path file_path, const ObjectFileInfo&
+    // placeholder in DeduceDebugSymbolMissingSizesAsDistanceFromNextSymbol.
+    symbol_info.set_size(ObjectFile::kUnknownSymbolSize);
+    symbol_infos->emplace_back(std::move(symbol_info));
+  }
+}
+
+}  // namespace
+
+PdbFileLlvm::PdbFileLlvm(std::filesystem::path file_path, const ObjectFileInfo& object_file_info,
+                         std::unique_ptr<llvm::pdb::IPDBSession> session)
+    : file_path_(std::move(file_path)),
+      object_file_info_(object_file_info),
+      session_(std::move(session)) {}
+
+[[nodiscard]] ErrorMessageOr<orbit_grpc_protos::ModuleSymbols> PdbFileLlvm::LoadDebugSymbols() {
+  std::vector<SymbolInfo> symbol_infos;
+  absl::flat_hash_set<uint64_t> address_from_module_debug_stream;
+
+  auto* native_session = dynamic_cast<llvm::pdb::NativeSession*>(session_.get());
+  ORBIT_CHECK(native_session != nullptr);
+  llvm::pdb::PDBFile& pdb_file = native_session->getPDBFile();
+
+  if (!pdb_file.hasPDBDbiStream()) {
+    return ErrorMessage("PDB file does not have a DBI stream.");
+  }
+
+  llvm::Expected<llvm::pdb::DbiStream&> debug_info_stream = pdb_file.getPDBDbiStream();
+  // Given that we check hasPDBDbiStream above, we must get a DbiStream here.
+  ORBIT_CHECK(debug_info_stream);
+
+  if (!pdb_file.hasPDBTpiStream()) {
+    return ErrorMessage("PDB file does not have a TPI stream.");
+  }
+  llvm::Expected<llvm::pdb::TpiStream&> type_info_stream = pdb_file.getPDBTpiStream();
+  // Given that we check hasPDBTpiStream above, we must get a TpiStream here.
+  ORBIT_CHECK(type_info_stream);
+
+  llvm::FixedStreamArray<llvm::object::coff_section> section_headers =
+      debug_info_stream->getSectionHeaders();
+
+  OUTCOME_TRY(LoadDebugSymbolsFromModuleStreams(
+      pdb_file, debug_info_stream.get(), type_info_stream.get(), section_headers, object_file_info_,
+      &symbol_infos, &address_from_module_debug_stream));
+
+  if (!pdb_file.hasPDBPublicsStream()) {
+    return ErrorMessage("PDB file does not have a public symbol stream.");
+  }
+  llvm::Expected<llvm::pdb::PublicsStream&> public_symbol_stream = pdb_file.getPDBPublicsStream();
+  ORBIT_CHECK(public_symbol_stream);
+
+  if (!pdb_file.hasPDBSymbolStream()) {
+    return ErrorMessage("PDB file does not have a symbol stream.");
+  }
+  llvm::Expected<llvm::pdb::SymbolStream&> symbol_stream = pdb_file.getPDBSymbolStream();
+  ORBIT_CHECK(symbol_stream);
+
+  LoadDebugSymbolsFromPublicSymbolStream(public_symbol_stream.get(), symbol_stream.get(),
+                                         section_headers, object_file_info_, &symbol_infos,
+                                         &address_from_module_debug_stream);
+
+  DeduceDebugSymbolMissingSizesAsDistanceFromNextSymbol(&symbol_infos);
+
+  if (symbol_infos.empty()) {
+    return ModuleSymbols{};
+  }
+
+  // The `DeduceDebugSymbolMissingSizesAsDistanceFromNextSymbol` can't fix the size of the last
+  // symbol (if that symbol is missing the size). But we can try and check if we are lucky and find
```

Comment:
Can you define what being lucky means? Why are there cases when we find the size, and others when we don't?